apps/glusterfs/app_device.go

```diff
@@ -350,3 +350,92 @@ func (a *App) DeviceSetState(w http.ResponseWriter, r *http.Request) {
 		return "", nil
 	})
 }
+
+func (a *App) DeviceResync(w http.ResponseWriter, r *http.Request) {
+
+	vars := mux.Vars(r)
+	deviceId := vars["id"]
+
+	var (
+		device *DeviceEntry
+		node   *NodeEntry
+	)
+
+	// Get device info from DB
+	err := a.db.View(func(tx *bolt.Tx) error {
+		var err error
+		device, err = NewDeviceEntryFromId(tx, deviceId)
+		if err != nil {
+			return err
+		}
+		node, err = NewNodeEntryFromId(tx, device.NodeId)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err == ErrNotFound {
+		http.Error(w, err.Error(), http.StatusNotFound)
+		return
+	} else if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		logger.Err(err)
+		return
+	}
+
+	logger.Info("Checking for device %v changes", deviceId)
+
+	// Check and update device in background
+	a.asyncManager.AsyncHttpRedirectFunc(w, r, func() (seeOtherUrl string, e error) {
+
+		// Get actual device info from manage host
+		info, err := a.executor.GetDeviceInfo(node.ManageHostName(), device.Info.Name, device.Info.Id)
+		if err != nil {
+			return "", err
+		}
+
+		if device.Info.Storage.Free == info.Size {
+			logger.Info("Device %v is up to date", device.Info.Id)
+			return "", nil
+		}
+
+		logger.Debug("Free space of '%v' (%v) has changed %v -> %v", device.Info.Name, device.Info.Id,
+			device.Info.Storage.Free, info.Size)
+
+		// Update device
+		err = a.db.Update(func(tx *bolt.Tx) error {
+
+			// Reload device in current transaction
+			device, err := NewDeviceEntryFromId(tx, deviceId)
+			if err != nil {
+				logger.Err(err)
+				return err
+			}
+
+			newFreeSize := info.Size
+			newTotalSize := newFreeSize + device.Info.Storage.Used
```

Comment:
The shrinking is totally fine and it is actually one of use cases. It could be useful even without any changes of the disk size. Let's consider that one copied a big file to the disk (directly by using filesystem) then heketi's disk free size is not correct anymore and resync will fix it. 
Total disk size in heketi in this case will be less then before, the free size will reflect actual free size left on the disk and the used size won't change.